Added Ctrl-L to FSI

voronoipotato opened the pull request:
I added this feature to FSI. I also cleared up a bit of the code where it was using character codes instead of keys.

dsyme commented:
@voronoipotato Could you jot down what manual testing you have done on this? Thanks

vasily-kirichenko commented on src/fsharp/fsi/console.fs:
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
+            if previous = "" then
+                setPrompt(true)
+            else
+                setPrompt(previous.[previous.Length - 2..previous.Length - 1] = ";;")            
```
Is this always safe?

voronoipotato replied:
fixed with EndsWith

voronoipotato commented:
Not a ton don, I just used it normally hitting ctrl-l while using fsi making sure it respected multi line, that it didn't nuke my history or my variables and functions already defined.

voronoipotato commented:
I think Vasily is correct it will crash if you have a single character as your previous command. I'll fix that and address Carter's nits

dsyme commented:
> Not a ton don, I just used it normally hitting ctrl-l while using fsi making sure it respected multi line, that it didn't nuke my history or my variables and functions already defined.

OK, could you run through all the key bindings implemented and make sure they all continue to work too, thanks

cartermp commented:
@voronoipotato Have you had a chance to think about addressing the feedback?

voronoipotato commented:
I have had a chance, sorry I've been busier than usual. I'll be able to work on it this weekend.

voronoipotato commented:
I tested the other commands, and they behaved the same as before. The only things of note I found is that Ctrl-A highlights all for me, and crtl-f initiates find for me, but it also does this on the previous version. I changed the string as array access and comparison for EndsWith which is cleaner and simpler.

cartermp commented:
@voronoipotato Have you had a chance to address feedback?

voronoipotato commented:
Hmm I thought that's what my change covered. I thought I had but even if so I definitely flopped on proper communication, many apologies. I'll review it this evening and make sure it addresses all the feedback.  Thanks for the ping @cartermp .

voronoipotato commented on src/fsharp/fsi/console.fs:
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
+                if prompt then //we only allow clearing if prompt is ">"
+                    Console.Clear()
+                    Console.Write ( x.Prompt)
+                    Console.Write( input.ToString())
```
removed extra space

voronoipotato commented on src/fsharp/fsi/console.fs:
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
+            let setPrompt(prompt) = 
+                if prompt then //we only allow clearing if prompt is ">"
+                    Console.Clear()
+                    Console.Write ( x.Prompt)
```
removed extra space

voronoipotato commented on src/fsharp/fsi/console.fs:
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
+        let clear() = 
+            current := input.Length
+            let setPrompt(prompt) = 
+                if prompt then //we only allow clearing if prompt is ">"
```
added space after //, capitalized W

voronoipotato commented on src/fsharp/fsi/console.fs:
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
+                setPrompt(true)
+            else
+                setPrompt(previous.[previous.Length - 2..previous.Length - 1] = ";;")            
+
```
removed trailing whitespace.

voronoipotato commented on src/fsharp/fsi/console.fs:
```diff
@@ -377,7 +378,25 @@ type internal ReadLineConsole() =
-        
+        let clear() = 
+            current := input.Length
+            let setPrompt(prompt) = 
```
`setPrompt(prompt)` was changed to `setPrompt prompt`.

voronoipotato commented:
Alright I've responded to all the feedback. The changes were done in https://github.com/dotnet/fsharp/pull/5808/commits/296f3f267f3a8977df39add7390c9fcd5a8d1d9a . Please let me know if you have any additional questions or if I've horribly misunderstood anything :) .

cartermp commented:
@voronoipotato Could you also resolve conflicts?

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=5808) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=5808) before we can accept your contribution.<br/><br/>:x: voronoipotato [sign now](https://cla.dotnetfoundation.org/dotnet/fsharp?pullRequest=5808)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/fsharp?pullRequest=5808) it.</sub>

voronoipotato commented:
@cartermp  it is done. Most of the conflicts were from me unconsciously adding semicolons after spending too much time in C#.

voronoipotato commented:
Linux CI failed with 

```
-> Could not download FSharpVSPowerTools.Core 2.3.
-> Couldn't get package details for package FSharpVSPowerTools.Core 2.3 on https://www.nuget.org/api/v2.
-> All possible sources are already blacklisted. 
```

cartermp commented:
@mmitche we get failures like this on the Linux agents all the time:

```
Paket failed with
-> Could not download FSharpVSPowerTools.Core 2.3.
-> Couldn't get package details for package FSharpVSPowerTools.Core 2.3 on https://www.nuget.org/api/v2.
-> All possible sources are already blacklisted. 
	 - 1_https://www.nuget.org/api/v2/Packages(Id='fsharpvspowertools.core',Version='2.3')
	 - 1_https://www.nuget.org/api/v2/Packages(Id='fsharpvspowertools.core',Version='2.3.0')
	 - 1_https://www.nuget.org/api/v2/Packages(Id='FSharpVSPowerTools.Core',Version='2.3.0')
	 - 1_https://www.nuget.org/api/v2/Packages(Id='FSharpVSPowerTools.Core',Version='2.3')
	 - 2_https://www.nuget.org/api/v2/Packages?$filter=(Id eq 'FSharpVSPowerTools.Core') and (NormalizedVersion eq '2.3.0')
	 - 2_https://www.nuget.org/api/v2/Packages?$filter=(tolower(Id) eq 'fsharpvspowertools.core') and (NormalizedVersion eq '2.3.0')
	 - 2_https://www.nuget.org/api/v2/Packages?$filter=(Id eq 'FSharpVSPowerTools.Core') and (Version eq '2.3.0')
	 - 2_https://www.nuget.org/api/v2/Packages?$filter=(tolower(Id) eq 'fsharpvspowertools.core') and (Version eq '2.3')
	 - 2_https://www.nuget.org/api/v2/Packages?$filter=(tolower(Id) eq 'fsharpvspowertools.core') and (Version eq '2.3.0')
```

It's not just this package, it could be a different one at a different time. Is this something we're aware of overall?

mmitche commented:
I haven't seen such an error.  @dotnet/dnceng does this ring a bell at all?

MattGal commented:
@mmitche it rings a bell but I've never seen a mono SSL implementation implicated as the culprit:

```
Source 'https://www.nuget.org/api/v2' exception: System.Exception: error ---> System.Exception: Could not retrieve data from 'https://www.nuget.org/api/v2/Packages?$filter=(tolower(Id) eq 'fsharpvspowertools.core') and (Version eq '2.3.0')' ---> System.Net.Http.HttpRequestException: An error occurred while sending the request ---> System.Net.WebException: Error: TrustFailure (Authentication failed, see inner exception.) ---> System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception. ---> Mono.Btls.MonoBtlsException: Ssl error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
  at /build/mono-5.20.1.19/external/boringssl/ssl/handshake_client.c:1132
```
... the exceptions are kind of similar to when we were using desktop .NET and Nuget moved to TLS 1.2 being mandatory.  That said, that issue was 100% repro so perhaps we need to ping some mono folks if it's intermittent.

voronoipotato commented:
@KevinRansom any reservations?

cartermp commented:
All green, @KevinRansom for review

cartermp commented:
Also adding @brettfo.

@KevinRansom is on vacation and may/may not get to things for the remainder of the month.

KevinRansom commented:
@voronoipotato,  thanks for doing this mate.  I'm sorry it dropped off my radar for such a long time.  I will try to do better next time.

Anyway ... thanks

Kevin

voronoipotato commented:
@KevinRansom it's cool, it was relatively low priority. Thanks for taking the time out so that it could be pulled through. 👍
